lib/asm510/generator.ex

```diff
@@ -135,22 +165,225 @@ defmodule ASM510.Generator do
           end
         end
 
+      {{:macro, macro_name, macro_args, macro_body}, _} ->
+        {:ok,
+         %State{
+           state
+           | macros: Map.put(state.macros, macro_name, {macro_name, macro_args, macro_body})
+         }}
+
+      {:exit_macro, line_number} ->
+        # A bit odd, but we need to bubble up until we hit a macro
+        {{:exit_macro, line_number}, state}
+
       {{:call, opcode, args}, line_number} ->
-        arg_values_result =
-          args
-          |> Enum.reduce_while({:ok, []}, fn {:expression, expr}, {:ok, args} ->
-            with {:ok, value} <- Expression.evaluate(expr, line_number, state.env) do
+        upcased_opcode = String.upcase(opcode)
+
+        cond do
+          Opcodes.is_opcode(upcased_opcode) ->
+            generate_opcode_call(upcased_opcode, args, line_number, state)
+
+          Map.has_key?(state.macros, opcode) ->
+            generate_macro_call(Map.get(state.macros, opcode), args, line_number, state)
+
+          true ->
+            {:error, line_number, {:unknown_opcode, opcode}}
+        end
+    end
+  end
+
+  defp generate_opcode_call(opcode, args, line_number, state) do
+    arg_values_result =
+      args
+      |> Enum.reduce_while({:ok, []}, fn arg, {:ok, args} ->
+        case arg do
+          nil ->
+            {:halt, {:error, line_number, {:missing_opcode_argument, opcode}}}
+
+          arg ->
+            with {:ok, value} <- get_arg_value(arg, line_number, state.env) do
               {:cont, {:ok, args ++ [value]}}
             else
               error -> {:halt, error}
             end
-          end)
+        end
+      end)
+
+    with {:ok, arg_values} <- arg_values_result,
+         {:ok, bytes} <- Opcodes.get_opcode(String.upcase(opcode), arg_values, line_number) do
+      bytes
+      |> Enum.reduce(state, &(&2 |> put_byte(&1) |> increment_pc()))
+      |> then(&{:ok, &1})
+    end
+  end
+
+  defp generate_macro_call(
+         {macro_name, macro_args, macro_body},
+         args,
+         macro_line_number,
+         state
+       ) do
+    if length(args) > length(macro_args) do
+      {:error, macro_line_number,
+       {:too_many_arguments, macro_name, length(macro_args), length(args)}}
+    else
+      define_macro_identifier = fn arg_name, name, env ->
+        env
+        |> Map.put("'#{arg_name}", name)
+        # Even if this identifier isn't defined yet, it might be
+        # defined during the course of the macro
+        |> Map.put("\\#{arg_name}", fn line, env ->
+          case Map.fetch(env, name) do
+            :error ->
+              {:error, line, {:undefined_symbol, name}}
+
+            result ->
+              result
+          end
+        end)
```

Comment:
Does this lazy evaluation allow you to reference labels defined after the macro?

Reply:
This lazy evaluation enables the following:
```
.macro define x
.set 'x, 1
.set 'x, \x + 1
.endm

define foo
```

Although `foo` is undefined when the macro is called, the macro doesn't refer to its value until after it's defined, at which point the lazy evaluation will succeed. This doesn't help if its value is defined later in the program though; only if it's defined between the start of the macro and when the macro gets its value.